src/liteclient.cc

```diff
@@ -93,50 +93,60 @@ LiteClient::LiteClient(Config& config_in, const AppEngine::Ptr& app_engine, bool
-    if (update_finalization_result.isSuccess()) {
-      LOG_INFO << "Marking target install complete for: " << pending_target->filename();
-      storage->saveInstalledVersion("", *pending_target, InstalledVersionUpdateMode::kCurrent);
-    } else if (data::ResultCode::Numeric::kInstallFailed == update_finalization_result.result_code.num_code) {
-      LOG_ERROR << "Failed to finalize the installation, rollback has happened";
-      // rollback has happenned, unset is_pending and was_installed flags of the given Target record in DB
-      storage->saveInstalledVersion("", *pending_target, InstalledVersionUpdateMode::kNone);
+data::InstallationResult LiteClient::finalizePendingUpdate(boost::optional<Uptane::Target>& target) {
+  data::InstallationResult ret{data::ResultCode::Numeric::kNeedCompletion, ""};
+  InstalledVersionUpdateMode mode = InstalledVersionUpdateMode::kPending;
+  bool rollback = false;
+  LOG_INFO << "Finalize Pending, Target: " << target->filename() << ", hash: " << target->sha256Hash();
+
+  ret = package_manager_->finalizeInstall(*target);
+  if (ret.isSuccess()) {
+    LOG_INFO << "Finalize Complete, Target: " << target->filename();
+    mode = InstalledVersionUpdateMode::kCurrent;
+  } else {
+    rollback = data::ResultCode::Numeric::kInstallFailed == ret.result_code.num_code;
+    // update dB Target record clearing flags (is_pending, was_installed)
+    if (rollback) {
+      LOG_ERROR << "Finalize Failed (rollback happened), Target" << target->filename();
+      mode = InstalledVersionUpdateMode::kNone;
     }
+  }
+  if (ret.isSuccess() || rollback) {
+    storage->saveInstalledVersion("", *target, mode);
+  }
+  return ret;
+}
+
+bool LiteClient::finalizeInstall() {
+  data::InstallationResult ret{data::ResultCode::Numeric::kNeedCompletion, ""};
+  boost::optional<Uptane::Target> pending;
+
+  storage->loadInstalledVersions("", nullptr, &pending);
+  if (!!pending) {
+    ret = finalizePendingUpdate(pending);
   } else {
-    LOG_INFO << "There is no any pending Target installation";
+    LOG_INFO << "No Pending Installs";
   }
 
-  const auto current_target = getCurrent();
-  update_request_headers(http_client, current_target, config.pacman);
-  writeCurrentTarget(current_target);
+  const auto current = getCurrent();
+  update_request_headers(http_client, current, config.pacman);
+  writeCurrentTarget(current);
 
-  if (data::ResultCode::Numeric::kNeedCompletion != update_finalization_result.result_code.num_code) {
-    // if finalization has happened, either succefully (kOk) or not (kInstallFailed)
-    // we send the installation finished report event.
-    notifyInstallFinished(*pending_target, update_finalization_result);
+  // finalize done, notify the backend
+  if (!ret.needCompletion()) {
+    notifyInstallFinished(*pending, ret);
   }
-  return data::ResultCode::Numeric::kOk == update_finalization_result.result_code.num_code;
+  return ret.isSuccess();
```

Comment:
@mike-sul I supposed this is equivalent?

Reply:
Not exactly, but in this context yes, since OstreeManager::finalizeInstall doesn't return kAlreadyProcessed. 
```
success(result_code_in.num_code == ResultCode::Numeric::kOk ||
                result_code_in.num_code == ResultCode::Numeric::kAlreadyProcessed),
```